{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Bash script  ##################################\n",
    "#copy files from talapas to a working directory\n",
    "cp /projects/bgmp/shared/deduper/*sam .\n",
    "\n",
    "#first sort files using samtools \n",
    "#general format:\n",
    "#       samtools sort [-l level] [-m maxMem] [-o out.bam] [-O format]\n",
    "#                 [-n] [-t tag] [-T tmpprefix] [-@ threads]\n",
    "#                 [in.sam|in.bam|in.cram]\n",
    "                 \n",
    "#no need to conver sam file to bam file\n",
    "#you can input sam, bam or cram files and outout whatever you want \n",
    "                \n",
    "#sort by QNAME first\n",
    "samtools sort -o sorted_dset1.sam -O sam -n Dataset1.sam \n",
    "\n",
    "#then sort by UMI\n",
    "samtools sort -o sorted_1.sam -O sam  sorted_dset1.sam \n",
    "\n",
    "########### Test files #####################\n",
    "#create a subset of your files to run your data \n",
    "#i will start with 5,000 lines can redirect it to a file called test1.sam\n",
    "less sorted_1.sam  | head -5000 > test1.sam\n",
    "\n",
    "#download this file to local computer so you can import it into jupyter notebook\n",
    "scp t1:/projects/bgmp/dchilin/Bi624/deduper/files/sorted/resorted/test_files/test* ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########### python script ###############################################################################################\n",
    "\n",
    "######## arg parse #####################################################################################################\n",
    "\n",
    "import argparse\n",
    "def get_arguments():\n",
    "    parser = argparse.ArgumentParser(description=\"Demultiplex\")\n",
    "    parser.add_argument('-f', '--input_file', help='insert absolute file path', required=True, type=str)\n",
    "    parser.add_argument('-o', '--output_file', help='insert absolute file path', required=False, type=str)\n",
    "    parser.add_argument(\"-p\", \"--paired_end\", help=\"designates file is paired end(not single-end)\", required=False, type=str)\n",
    "    parser.add_argument(\"-s\", \"--single_end\", help=\"designates file is single end(not paired-end)\", required=False, type=str)\n",
    "    parser.add_argument(\"-u\", \"--umi_list\", help=\"designates file containing the list of UMIs\", required=False, type=str)\n",
    "    return parser.parse_args()\n",
    "\n",
    "#code when using arg parse in bash:\n",
    "#$ ./script_name -f /path/filename -o other_arguments\n",
    "\n",
    "\n",
    "############## variables ###################################################################################################\n",
    "#variables needed when parsing the file\n",
    "args = get_arguments()\n",
    "input_file = args.input_file\n",
    "output_file = args.output_file\n",
    "umi_list = args.umi_list\n",
    "p = args.paired_end\n",
    "s = args.single_end\n",
    "\n",
    "#creat a variable to call your file \n",
    "#since we are using arg parse, you dont need to do this (just comment it out for now)\n",
    "#f = '/projects/bgmp/dchilin/Bi624/deduper/files/sorted/resorted/sorted_1.sam'\n",
    "\n",
    "#umi file (its just commented out for now)\n",
    "#u='/projects/bgmp/dchilin/Bi624/deduper/files/sorted/resorted/UMI.txt'\n",
    "\n",
    "#create/initiate empty variables to keep track of current & previous chromosome \n",
    "current_chr = ''\n",
    "previous_chr = ''\n",
    "\n",
    "########### Dictionaries ######################################################################################################\n",
    "\n",
    "#create dictionary to store all your real UMIs that were given to you \n",
    "#u is defined on top via arg parse\n",
    "\n",
    "UMI_dict={}\n",
    "for line in open(umi_list).readlines():\n",
    "    umi=line.strip()\n",
    "    UMI_dict[umi]=None\n",
    "\n",
    "\n",
    "#dictionary to check if there are any duplicates\n",
    "#key: QNAME, Chrom, and POS\n",
    "dup_dict = {}\n",
    "\n",
    "\n",
    "################ Functions ###########################################################################################\n",
    "\n",
    "#some functions require RegEx, so import that pacakge\n",
    "import re\n",
    "#isolating UMI from QNAME\n",
    "#example :  NS500451:154:HWKTMBGXX:1:11101:25533:1187:GTTCACCT\n",
    "#  UMI : GTTCACCT\n",
    "def isolate_UMI(string):\n",
    "    #UMI is found at the very end of the string, use $\n",
    "    UMI = re.findall(r'([A-Z]+$)', string)[0]\n",
    "    #re.findall returns all matches as a list. \n",
    "    #return [0], even thouhg there is only one things in the list \n",
    "    return (UMI)\n",
    "\n",
    "\n",
    "#checking for single/paired ends\n",
    "#bitwise flag 1 means strand is paired-end\n",
    "def check_read(flag):\n",
    "    read= ((int(flag )& 1) == 1)\n",
    "    #if True: paired-end (flag equals 1)\n",
    "    #if False: single-end (flag does not equal 1)\n",
    "    return read\n",
    "           \n",
    "#check if strand is forward or reverse \n",
    "#bitwise flag 16 is reverse strand \n",
    "def check_strand_read(flag):\n",
    "    strand = ((int(flag) & 16) == 16)\n",
    "    #if True: Reverse Strand (flag equals 16)\n",
    "    #if False: Forward Strand (flag does not equal 16)\n",
    "    return strand     \n",
    "    \n",
    "     \n",
    "#check cigar string for forward strand ONLY\n",
    "def check_cigar_forward_string(string):\n",
    "    #position start at the left most end\n",
    "    #softclipping can only occur on either ends of the string\n",
    "    #only account for S thats on the left side of the string\n",
    "    #ignore any S found on the the right side of the string (it will not affect the position)\n",
    "    #N, I, & D can only happen in the middle of the read, so dont worry too much about these here \n",
    "        #it will afect the position\n",
    "    #find S's before any M's or any other letter\n",
    "    S = re.findall(r'(^\\d+[S])', string)\n",
    "    #isolate the number before S\n",
    "    S = sum([ int(i.strip('S'))  for i in S])\n",
    "    return S\n",
    "\n",
    "\n",
    "def check_cigar_reverse_string(string):\n",
    "    #position start on left most of the string, however this is a reverse string \n",
    "    #you want to adjust position to were read is ending, which is the right side\n",
    "    #find S's on the right side of the string.\n",
    "    #Ignore the S's on the left (they do not affect the position of the string)\n",
    "    #find M, D, & N\n",
    "    S = re.findall(r'(\\d+[S]$)', string)\n",
    "    M = re.findall(r'(\\d+[M])', string)\n",
    "    D = re.findall(r'(\\d+[D])', string)\n",
    "    N = re.findall(r'(\\d+[N])', string)\n",
    "    #isolate the numbers\n",
    "    S = sum([ int(i.strip('S'))  for i in S])\n",
    "    M = sum([ int(i.strip('M'))  for i in M])\n",
    "    D = sum([ int(i.strip('D'))  for i in D])\n",
    "    N = sum([ int(i.strip('N'))  for i in N])\n",
    "    #sum S, M, D, N\n",
    "    total = (S + M + D + N)\n",
    "    return total\n",
    "    #note: we do not account for any insertions (I) because it does not consume(affect) the reference read\n",
    "   \n",
    "\n",
    "########### line counters #####################################################################################\n",
    "#line loop is going through \n",
    "#we just need this for the first line the script is reading through to set previuos_chr to something)\n",
    "ctr = 1\n",
    "\n",
    "#for undetermined/bad UMI's\n",
    "bad_umi = 0\n",
    "\n",
    "#for paired_end reads\n",
    "paired_reads = 0\n",
    "\n",
    "#for reverse reads\n",
    "reverse_read = 0\n",
    "\n",
    "#for forwards reads \n",
    "forward_read = 0\n",
    "\n",
    "#single reads we want to keep\n",
    "single_reads = 0\n",
    "\n",
    "#for duplicated reads we dont want to keep\n",
    "duplicated_reads = 0\n",
    "\n",
    "################ code ##############################################################################################\n",
    "#open same file\n",
    "#[:-4] is used to removed the last 4 characters in the file (.sam)\n",
    "with open (input_file, 'r') as fh, \\\n",
    "    open(output_file + \"_deduped.sam\", \"w+\") as wh, \\\n",
    "    open(output_file + '_unwanted_duped_reads.sam', 'w+') as d_reads, \\\n",
    "    open(output_file + '_undertermined_UMIs.txt', 'w+') as b_umi:\n",
    "    i = 0\n",
    "    for line in fh:\n",
    "        i += 1\n",
    "        #ignore header lines (start with @)\n",
    "        #spit those out to deduped file\n",
    "        if line.startswith('@'):\n",
    "            wh.write(line)\n",
    "        else:\n",
    "            #separate columns from each line into indexes by tab delimination\n",
    "            #remember index starts with 0\n",
    "            column = line.split('\\t')\n",
    "            #create variables for each index\n",
    "            # we only want QNAME, FLAG, RNAME, POS, and CIGAR\n",
    "            #the others are not needed, but i just labeled it for future reference\n",
    "            QNAME = column[0] #UMI\n",
    "            FLAG = column[1] #single/pair, +/-\n",
    "            RNAME = column[2] #chromosome\n",
    "            POS = column[3] #position\n",
    "            MAPQ = column[4]\n",
    "            CIGAR = column[5] #soft clipping\n",
    "            RNEXT = column[6]\n",
    "            PNEXT = column[7]\n",
    "            TLEN = column[8]\n",
    "            SEQ = column[9]\n",
    "            QUAL = column[10]\n",
    "\n",
    "            ### memory efficiency ##################################################################################\n",
    "            \n",
    "            #this is for memory efficiency\n",
    "            #idea is go through all of one chromosome, once you reach a different chromosome, empty the dictionary\n",
    "            #set current chromosome to RNAME (index that has the chromosome)\n",
    "            current_chr = RNAME\n",
    "            #set previous chromose to the current chromosome\n",
    "            #this only happens with the very first line reaches this part of the code\n",
    "            #this is a way to start current and previous chromosome the first chromoomse on the file\n",
    "            if ctr == 1: \n",
    "                previous_chr = current_chr\n",
    "            ctr += 1\n",
    "            #once you go through one chromosome, and reach a new chromosome we want to empty the dictioanry\n",
    "            if current_chr != previous_chr:\n",
    "                #if current chromosome does not equal previous chromosome(meaning we have encounter a new chromosome)\n",
    "                #we will set our previous chromosome to the current one\n",
    "                previous_chr = current_chr\n",
    "                #emtpy dictionary\n",
    "                dup_dict = {}\n",
    "\n",
    "            ### UMI validation ########################################################################################  \n",
    "            \n",
    "            #isolate the UMI from QNAME  using the function for it\n",
    "            #then verify if UMI exists in the UMI dictionary \n",
    "            UMI = isolate_UMI(QNAME)\n",
    "            if UMI not in UMI_dict:\n",
    "                #UMI not in dictionary, its no good\n",
    "                #ignore this UMI, throw it to a dump file, and keep track of bad umis\n",
    "                b_umi.write(line)\n",
    "                bad_umi += 1\n",
    "            else:\n",
    "                 #UMI exists in the dicitonary, continue with code\n",
    "\n",
    "                ### Paired vs Single-end reads #########################################################################\n",
    "                \n",
    "                #check if read is paired-end/single end\n",
    "                if check_read(FLAG) == True:\n",
    "                #if true, this is paired-end read\n",
    "                #code cant handle paired reads, print an error message and quit\n",
    "                    print('ERROR: Paired-end read')\n",
    "                    paired_reads += 1\n",
    "                    break\n",
    "                else:\n",
    "                #if not true, this is single-end read, continue with code\n",
    "\n",
    "                    ###Forward/Reverse strands #########################################################################\n",
    "                    ### (adjusting postion if needed ) #################################################################\n",
    "\n",
    "                    #check if strand is Forward/Reverse\n",
    "                    if check_strand_read(FLAG) == True:\n",
    "                    #if FLAG is 16, its a reverse strand\n",
    "                        reverse_read += 1\n",
    "                        cigar = check_cigar_reverse_string(CIGAR)\n",
    "                        #adjPOS = int(POS) - 1\n",
    "                        adjPOS = int(POS) + int(cigar) - 1\n",
    "                        #add '-' to adjPOS to differentiate from the forward strand in the dictionary \n",
    "                        adjPOS = str(adjPOS) + '-'\n",
    "                    else:\n",
    "                    #if not 16, its a forward strand\n",
    "                        forward_read += 1\n",
    "                        if 'S' in CIGAR:\n",
    "                            S = check_cigar_forward_string(CIGAR)\n",
    "                            #subtract the S from POS so that it can adjust for those numbers that were \"clipped out\"\n",
    "                            adjPOS = int(POS) - int(S)\n",
    "                            #indicate adjPOS that this is a forward(+) strand (to help differentiate from the reverse strand )\n",
    "                            adjPOS = str(adjPOS) + '+'\n",
    "                        else:\n",
    "                            #set POS equal to a variable so we can call on the cigars that dont have S's\n",
    "                            adjPOS = str(POS) + '+'\n",
    "\n",
    "                    ### duplicate verfication ###################################################################################\n",
    "\n",
    "                    #all positions should be adjusted\n",
    "                    #add specific columns to dictionary to check for duplicates\n",
    "                    #Only interested if UMI, flag, chromosome and position matches\n",
    "                    #separate it by commas and '' so its easier to read\n",
    "                    #key = (str(UMI))+','+\"'\"+str(strand)+','+\"'\"+str(RNAME)+\"'\"+','+\"'\"+str(adjPOS)+\"'\" \n",
    "                    #alternatie quicker format than above\n",
    "                    key = '{},{},{}'.format(UMI, RNAME, str(adjPOS))\n",
    "                    #check if there are any duplicate reads in the duplicate dictionary\n",
    "                    if key not in dup_dict:\n",
    "                        #not in dictionary, add it to the dictionary\n",
    "                        dup_dict[key] = 1\n",
    "                        #for memory efficiency, output this line to the deduper file (this is the file with all the wanted silge reads )\n",
    "                        wh.write(line)\n",
    "                        #increment counter\n",
    "                        single_reads +=1\n",
    "                    else:\n",
    "                        #keys are in the dictionary, there is a mathc. Get rid of this line.\n",
    "                        #we only want to keep one copy of this read.\n",
    "                        #throw out this read to the unwanted dupliated reads \n",
    "                        d_reads.write(line)\t\n",
    "                        #increment counter \n",
    "                        duplicated_reads +=1\n",
    "\n",
    "print('undertermined umi:', bad_umi) \n",
    "print('paired-end reads:', paired_reads)\n",
    "print('forward reads:', forward_read)                   \n",
    "print('reverse reads:', reverse_read)\n",
    "print('unwanted duplicates:', duplicated_reads)\n",
    "print('wanted deduped reads:', single_reads)\n",
    "\n",
    "\n",
    "### out put ######################################################################################################################\n",
    "\n",
    "#***note : \twanted + unwanted + undetermined umis + header = total lines in original file\n",
    "\n",
    "#dataset1: \n",
    "#lines in original file :1,013,204\n",
    "#lines in sorted file : 1,013,204\n",
    "#header(lines with @) : 24\n",
    "\t# undertermined umi: 8,190\n",
    "\t# paired-end reads: 0\n",
    "\t# forward reads: 1,004,945\n",
    "\t# unwanted duplicates: 369,005\n",
    "\t# wanted deduped reads: 635,985\n",
    "\n",
    "#dataset 2 : \n",
    "#lines in orignal file : 1,382,133\n",
    "#lines in sorted file : 1,382,133\n",
    "#lines wit#header(lines with @) : 24\n",
    "\t# undertermined umi: 9,319\n",
    "\t# paired-end reads: 0\n",
    "\t# reverse reads: 53\n",
    "\t# unwanted duplicates: 628,602\n",
    "\t# wanted deduped reads: 744,188\t\n",
    "\n",
    "#dataset 3 :\n",
    "#lines in orignial file: 5,721,174\n",
    "##lines in sorted file: 5,721,174\n",
    "#header(lines with @) : 24\n",
    "\t# undertermined umi: 48,939\n",
    "\t# paired-end reads: 0\n",
    "\t# reverse reads: 13,876\n",
    "\t# unwanted duplicates: 1,618,567\n",
    "\t# wanted deduped reads: 4,053,644"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test file ###############################################################################################################################################################################################################################################################################################################\n",
    "\n",
    "@SQ     SN:1    LN:195471971\n",
    "@SQ     SN:2    LN:182113224\n",
    "@SQ     SN:3    LN:160039680\n",
    "@SQ     SN:4    LN:156508116\n",
    "@SQ     SN:5    LN:151834684\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t2\t10\t36\t50M40S\t[10]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:25533:1187:GTTCACCT\t16\t2\t10\t36\t100M\t[11]\t0\t0\tCTTGGTAACTTTCAGAGAATTAGTCACAACTTCTGAAGCAACCACAGTCCATGCAAGTCGACTGGTTTCTC 6AEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEAEEEEEEE<EEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:25533:1187:GTTCACCT\t16\t2\t10\t36\t90M10S\t[13]\t0\t0\tCTTGGTAACTTTCAGAGAATTAGTCACAACTTCTGAAGCAACCACAGTCCATGCAAGTCGACTGGTTTCTC 6AEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEAEEEEEEE<EEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:25533:1187:GTTCACCT\t16\t2\t10\t36\t10S10M10N20D10I50M\t[14]\t0\t0\tCTTGGTAACTTTCAGAGAATTAGTCACAACTTCTGAAGCAACCACAGTCCATGCAAGTCGACTGGTTTCTC 6AEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEAEEEEEEE<EEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:25533:1187:TTCGCCTA\t16\t2\t10\t36\t100M\t[15]\t0\t0\tCTTGGTAACTTTCAGAGAATTAGTCACAACTTCTGAAGCAACCACAGTCCATGCAAGTCGACTGGTTTCTC 6AEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEAEEEEEEE<EEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:25533:1187:GTTCACCT\t16\t2\t20\t36\t10S90M\t[12]\t0\t0\tCTTGGTAACTTTCAGAGAATTAGTCACAACTTCTGAAGCAACCACAGTCCATGCAAGTCGACTGGTTTCTC 6AEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEAEEEEEEE<EEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:ATTTTTTT\t0\t2\t76814284\t36\t71M\t[1]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTNCAC\t0\t2\t76814284\t36\t71M\t[2]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTNCAC\t0\t2\t76814284\t36\t71M\t[17]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t16\t3\t10\t36\t100M\t[16]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t3\t10\t36\t100M\t[4]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t3\t10\t36\t50M40S\t[5]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:6251:1098:ATCCATGG\t0\t3\t10\t36\t60M10D10M\t[9]\t0\t0\tGGCGTTCCAAACCACGGTCATCTCTTCTTTGCTTACTTTAGTGACTTCTGGAGGATCAGGGCGGCCAGGTC /<EEAEEEEEEEEAEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t3\t60\t36\t50S50M\t[6]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t3\t60\t36\t50S10M10N30M\t[7]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:24260:1121:CTGTTCAC\t0\t3\t60\t36\t50S10M20I10D10M\t[8]\t0\t0\tTCCACCACAATCTTACCATCCTTCCTCCAGACCACATCGCGTTCTTTGTTCAACTCACAGCTCAAGTACAA 6AEEEEEEAEEAEEEEAAEEEEEEEEEAEEAEEAAEE<EEEEEEEEEAEEEEEEEAAEEAAAEAEEAEAE/ MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "NS500451:154:HWKTMBGXX:1:11101:6251:1098:AAAAAAAA\t0\t3\t76765947\t36\t71M\t[3]\t0\t0\tGGCGTTCCAAACCACGGTCATCTCTTCTTTGCTTACTTTAGTGACTTCTGGAGGATCAGGGCGGCCAGGTC /<EEAEEEEEEEEAEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE MD:Z:71 NH:i:1  HI:i:1  NM:i:0  SM:i:36 XQ:i:40 X2:i:0  XO:Z:UU\n",
    "                                                                                            \n",
    "                                                                                            \n",
    "expected output : \n",
    "    undertermined umi: 4\n",
    "    paired-end reads: 1\n",
    "    forward reads: 6\n",
    "    reverse reads: 6\n",
    "    unwanted duplicates: 5\n",
    "    wanted deduped reads: 8\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
